Configurable training

saranggalada opened the pull request:
1. Added two new scenarios: BraTS and CIFAR-10
2. Pytrain now allows for highly flexible and secure training pipeline configurations
3. Fixed bugs in training code and ACI deployment
4. Restructured scenario subdirectories to demarcate different parties and components

saranggalada committed:
Added LLM-finetune scenario local execution

saranggalada committed:
Added MRI-Segmentation scenario local execution

saranggalada committed:
added aci deployment steps

saranggalada committed:
Update load_base_model.py

kapilvgit committed:
build scripts

kapilvgit committed:
script permissions

kapilvgit committed:
support PrivateTrainVision in policy

saranggalada committed:
Update pipeline_config.json

saranggalada committed:
standardized config param names

kapilvgit committed:
naming changes

kapilvgit committed:
epsilon

saranggalada committed:
bug fixes and automation

saranggalada committed:
unified pytrain across scenarios and added bring your custom code functionality for TDCs

saranggalada committed:
fixed file naming and path handling

saranggalada committed:
introduced training customization wrappers and configurations. Added Cifar10 scenario

saranggalada committed:
Updated Readmes

saranggalada committed:
Updated Readmes

saranggalada committed:
security fixes to prevent malicious code injection and unapproved modules

saranggalada committed:
updated scenarios table

kapilvgit commented on README.md:
````diff
@@ -43,17 +44,22 @@ This scripts build the following containers.
 Alternatively, you can use pre-built container images from the ispirt repository by setting the following environment variable. Docker hub has started throttling which may effect the upload/download time, especially when images are bigger size. So, It is advisable to use other container registries, we are using azure container registry as shown below
 ```bash
-export CONTAINER_REGISTRY=ispirt.azurecr.io
+export CONTAINER_REGISTRY=depatraindevacr.azurecr.io
````
suggest we use the ispirt repo

kapilvgit commented on ci/Dockerfile.train:
```diff
@@ -1,23 +1,33 @@
+RUN pip3 --default-timeout=1000 install pyspark pandas opacus==1.5.3 onnx onnx2pytorch scikit-learn scipy matplotlib 
+RUN pip3 install safetensors h5py pyarrow
+
+## ADDED: For language tasks
```
Remove ADDED

kapilvgit commented on scenarios/brats/config/consolidate_pipeline.sh:
At some point we should consider a framework like Hydra for all configuration

kapilvgit commented on scenarios/brats/config/model_config_old.json:
Do we still need these files?

saranggalada replied:
Oops, forgot to delete

kapilvgit commented on scenarios/brats/deployment/azure/deploy.sh:
```diff
@@ -0,0 +1,173 @@
+      export BEARER_TOKEN=$(az account get-access-token --resource https://managedhsm.azure.net | jq -r .accessToken)
+      export AZURE_AKV_KEY_TYPE="oct-HSM"
+  fi
+
```
We should try and clean this up using a loop to go over all datasets and key combinations

saranggalada replied:
Yes, in a future PR we would ideally create a "Create your own scenario" guide, as part of which the repeated code for deployment is auto-generated from template. This would remove all the hardcoding we have done within these scripts, and factor in number of datasets, etc

kapilvgit commented on scenarios/cifar10/config/model_config.json:
for this scenario, can we not use ONNX?

saranggalada replied:
Yes, we could. Since it is very similar to the MNIST scenario, I thought we could demonstrate some variability/flexibility by changing model and dataset formats and showing that they work smoothly.

saranggalada committed:
Added XGBoost support and Credit-Risk scenario

saranggalada commented:
Comments addressed and PR updated.

**Summary of changes**:
1) Introduced xgb_train.py workflow for XGBoost training and renamed existing train.py to dl_train.py to remove ambiguity.
2) New scenario "Credit Risk" that joins 6 datasets from 4 TDPs and trains using XGBoost with DP. Here, one of the TDPs (Bank A) also acts as the TDC, to reflect a more realistic scenario of collaboration.
3) Moved parsers/contructors for model, dataset, eval and loss configs into utilities
4) Minor change in policy.rego to allow single TDC container (if base model is being instantiated inside CCR)
5) Included 2 alternate model_config samples for COVID, one that trains using XGBoost, another that packages model as a SafeTensor file and not the default ONNX.
6) Updated ReadMes: i) Added mermaid diagrams to illustrate config creation for training. ii) Added info about datasets used in scenarios. iii) Modified main readme scenarios table. iv) Renamed ACR and Contract Service identifiers.
7) Added parameter "provider" in Join config to differentiate from dataset name, since single provider may have more than one dataset

**To be done (in subsequent PRs)**:
1) "Create your own Scenario" guide and boilerplate automation
2) Classical ML training support
3) LLM Finetuning support
4) iSPIRT Contract Service 24*7 active demo, to provide users pre-signed contract sequence numbers for quick demos.
5) Optimize configurations / adopt existing standards (eg. Hydra / KubeFlow)

saranggalada committed:
Update README.md

saranggalada committed:
Update ci-build.yml

Updated local deployment path

saranggalada committed:
Update ci-local.yml

saranggalada committed:
updated ci
